### ISSUE TITLE
V8JS Segfaults with React DOM Server > 16.6.3 because of `Uint16Array`

### ISSUE THREAD

**Issue by virgofx** [phpv8/v8js#393]
In versions of React > 16.6.3 they start using the `Uint16Array`. This construct works fine when executing string; however, segfaults when creating snapshot:
```php
use V8Js;

$jsTest = 'var newArray = new Uint16Array(15);';

$v8js = new V8Js('PHP');
$v8js->executeString($jsTest); // WORKS

$result = V8Js::createSnapshot($jsTest); // Segfaults;
```

Maybe this is fixed in the v8 upstream and we need a new version? Otherwise, unable to get debugging setup correct to determine more information regarding segfault.

---

*php --version*
```
PHP 7.2.14-1+ubuntu16.04.1+deb.sury.org+1 (cli) (built: Jan 13 2019 10:05:18) ( NTS )
Copyright (c) 1997-2018 The PHP Group
Zend Engine v3.2.0, Copyright (c) 1998-2018 Zend Technologies
```

*php --ri v8js*
```
v8js

V8 Javascript Engine => enabled
V8 Engine Compiled Version => 7.1.302.12
V8 Engine Linked Version => 7.1.302.12
Version => 2.1.0
```

**Comment by virgofx** [comment]
@stesie Let me know if you're able to confirm and have any ideas

**Comment by timothympace** [comment]
I am diving into v8js and v8 and ran into this same issue. I noticed that the `mksnapshot` tool that v8 provides has two arguments vs `createSnapshot` only having one. They both use `embed_script` as an argument, but `mksnapshot` also provides a `warmup_script` parameter. Now, I don't fully understand why, but when I split library code vs execution code and pass the library code as `embed_script` and the execution code as `warmup_script` everything works great. I found the commit where they changed this: https://github.com/v8/v8/commit/f507bc1c7242d338b04d6bba5ea78cbb42cf0e13

I had compiled v8 in debug mode a few days ago and found that mksnapshot was failing because it was missing some callback. It fails this debug check here: https://github.com/v8/v8/blob/master/src/snapshot/partial-serializer.cc#L146

**Comment by virgofx** [comment]
@timothympace Awesome investigation analysis. Maybe @stesie can confirm. Would love to get this fixed to be able to use the latest React versions integrated with V8JS.

PS- It took me forever to figure out the error with `V8JS::createSnapshot()` because it segfault'd and worked fine executing string.  (Had to slice and dice code 50% of the time)

Again -- totally not needed but wondering if there is a better way or not sure if there is another method we could add somewhere to help with snapshot failures?

**Comment by timothympace** [comment]
@virgofx  Funny, we were doing the same thing. I spent Saturday morning eliminating pieces of code from react-dom/server until I isolated the issue. I came to the same conclusion: trying to create a snapshot from code that allocates any typed array will cause a segfault.

I said in my previous comment that `mksnapshot` will not segfault if you split the scripts into `embed_script` and `warmup_script`. I realize now that it's because allocations/side effects are not saved when using a warmup script (see https://github.com/v8/v8/commit/f507bc1c7242d338b04d6bba5ea78cbb42cf0e13), and that putting the calls to react-dom/server in the warmup_script doesn't really solve anything.

My tests went like this...

File: `test.js`:
```
function getTypedArray(size) {
    return new Uint16Array(size);
}

var test = getTypedArray(16);
```

File: `test2.js`:
```
function testWarmFunction() {
    return 'I am warm';
}

const typedArray = getTypedArray(16);
```

1st test:
```
./mksnapshot --startup-blob=snapshot.blob test.js test.2js
> Loading script for embedding: test.js
> Segmentation fault
```
Typed array allocation causes segfault.

2nd test (commenting out the call to getTypedArray in test.js):
```
./mksnapshot --startup-blob=snapshot.blob test.js test.2js
> Loading script for embedding: test.js
> Loading script for warm up: test2.js
```
This test succeeds because the embedded script avoids the typed array allocation. Even though the warmup script also creates an allocation, it doesn't segfault because the warmup script is just used to decide which functions get compiled into memory (further proof in test 3)

3rd test
```
root@app:/usr/local/src/v8/out/x64.release# php7.0 -a
Interactive mode enabled

php > $v8 = new V8Js('PHP', [], [], true, file_get_contents(__DIR__ . '/snapshot.blob'));
php > $v8->executeString('print(testWarmFunction());');
PHP Warning:  Uncaught V8JsScriptException: V8Js::compileString():1: ReferenceError: testWarmFunction is not defined in php shell code:1
Stack trace:
#0 php shell code(1): V8Js->executeString('print(testWarmF...')
#1 {main}
  thrown in php shell code on line 1
```
This shows that things defined in the warmup script are in fact not saved into the blob. This is why the allocation doesn't cause a segfault.

tl:dr:
1. embed_script should be used for all the code that should be parsed and executed and saved into the blob.
2. warmup_script should be used for deciding which functions get compiled into memory.
3. typed array allocations will cause a segfault when trying to serialize it into a snapshot.

Since I did all these tests with mksnapshot, I think its safe to say this issue is not coming from v8js. I am going to try compiling the latest version of v8 to see if this issue still exists.

Also, if you build for production the issue with react dom server goes away.  The typed array allocations only happen in the development build 😄

**Comment by timothympace** [comment]
They have a test written around this exact thing.. but it shortcuts a piece of code (no clue where that code is) that is responsible for passing some sort of callback.

Here is the test:
https://github.com/v8/v8/blob/dd732b4fb09ba3f2a2d13606c6a049f208f9a284/test/cctest/test-serialize.cc#L877
Here is where the callback function gets passed:
https://github.com/v8/v8/blob/dd732b4fb09ba3f2a2d13606c6a049f208f9a284/test/cctest/test-serialize.cc#L847

The callback function not being null is what causes the segfault: https://github.com/v8/v8/blob/master/src/snapshot/partial-serializer.cc#L146

**Comment by timothympace** [comment]
Don't know exactly when it was fixed, but the issue is gone in the master branch of v8.

**Comment by virgofx** [comment]
Just to clarify -- you're saying if we use the latest version in master/V8 it should work with no V8JS changes?

**Comment by timothympace** [comment]
I am saying that master/V8 has the issue fixed as far as using `mksnapshot` is concerned. I tried compiling v8js against it and ran into some failures. I made some small code changes by looking at the git blame to see how the API changed. I can open a pull request shortly.

**Comment by timothympace** [comment]
I take that back. My first test worked just fine, but other tests are not working. Going to investigate more.

**Comment by timothympace** [comment]
Previously, just instantiating and not assigning would throw the segfault: `new Uint16Array(16);`. That works fine on master now... but the second you try to assign it to a variable, it still segfaults. I'll still add my PR for updating V8JS to support the latest master branch though.

Thinking it's time to file an issue with v8.

**Comment by timothympace** [comment]
Here is the issue I just filed with v8: https://bugs.chromium.org/p/v8/issues/detail?id=8822
PR for building with master: https://github.com/phpv8/v8js/pull/395

**Comment by timothympace** [comment]
Forgot to post it sooner, but if anyone needs a workaround, I just posted it into the chromium bug tracker: https://bugs.chromium.org/p/v8/issues/detail?id=8822#c9

**Comment by timothympace** [comment]
Fix has been merged into V8: https://github.com/v8/v8/commit/290cd816518590acb7d5b5205eeb630434d1ad16

**Comment by virgofx** [comment]
@timothympace Do you know the next build that should contain the fix? Maybe at that time @stesie can rebuild the PPA binaries.

**Comment by timothympace** [comment]
Don't know what number it will be, but it looks like the next tagged release (so maybe 7.4.302?) will include it. It's available when building from the master branch though 😄

**Comment by virgofx** [comment]
Let me know if you confirm it's fixed in one of the latest builds.

**Comment by stesie** [comment]
Hej, some updates from my side (sorry for being so quite currently)

* I'm currently re-building the ubuntu packages based on 7.2.502.28 with the patch applied
* the warmup-script thing was introduced some time after adding snapshot support; and so far supporting it from php-v8js is on my wishlist, but not done yet

As far as I understand the difference, what you run in the actual startup script makes it directly into the heap snapshot. Using the warmup script can trigger JIT compilations in the startup code, which can again improve the startup codes performance ...

**Comment by virgofx** [comment]
@stesie  Looks like upstream fixed it https://chromium.googlesource.com/v8/v8/+log/refs/tags/7.5.1

Any chance we could rebuild the binaries in your PPA with this latest version so we can test?

**Comment by stesie** [comment]
@virgofx the 7.2 package has the fix applied.  I wanted to stick to the stable version line

**Comment by stesie** [comment]
See https://github.com/phpv8/ppa-packaging/compare/eb53033..5274d13

**Comment by virgofx** [comment]
could we add a 7.5 baseline also?

**Comment by stesie** [comment]
let me see, I'm pretty short on time recently ... but I'll see what I can do.  Regularly building the development branch (and otherwise it doesn't make much sense), is a good point to start fully automating things :)

Just out of curiosity, why do you want the 7.5 version line so much ... opposed to the "stable" 7.2 one?

**Comment by virgofx** [comment]
No preference, 7.2 is fine. And yes I agree , I want to automate the building 100%. I still have a workspace with all my development to get it 100% working in travis.

**Comment by virgofx** [comment]
Just updating ... the fix is in >= 7.5.1 which means we really need to get V8JS compiling against this version which is currently broken. I'll close this and reference the thread specifically related to getting V8JS working with 7.5

Ref: #414
